src/rules/noDuplicateVariableRule.ts

```diff
@@ -38,73 +39,56 @@ export class Rule extends Lint.Rules.AbstractRule {
-            this.handleSingleVariableIdentifier(node.name as ts.Identifier);
+function walk(ctx: Lint.WalkContext<void>): void {
+    const scopes: Array<Set<string>> = [new Set()];
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (utils.isFunctionScopeBoundary(node)) {
+            scopes.push(new Set());
+            ts.forEachChild(node, cb);
+            scopes.pop();
+            return;
         }
 
-        super.visitBindingElement(node);
-    }
-
-    public visitCatchClause(node: ts.CatchClause) {
-        // don't visit the catch clause variable declaration, just visit the block
-        // the catch clause variable declaration has its own special scoping rules
-        this.visitBlock(node.block);
-    }
-
-    public visitMethodSignature(_node: ts.SignatureDeclaration) {
-        // don't call super, we don't want to walk method signatures either
-    }
-
-    public visitTypeLiteral(_node: ts.TypeLiteralNode) {
-        // don't call super, we don't want to walk the inside of type nodes
-    }
-
-    public visitVariableDeclaration(node: ts.VariableDeclaration) {
-        const isSingleVariable = node.name.kind === ts.SyntaxKind.Identifier;
-
-        // destructuring is handled by this.visitBindingElement()
-        if (isSingleVariable && !Lint.isBlockScopedVariable(node)) {
-            this.handleSingleVariableIdentifier(node.name as ts.Identifier);
+        if (utils.isVariableDeclaration(node) && !utils.isBlockScopedVariableDeclaration(node)) {
+            const scope = scopes[scopes.length - 1];
+            forEachBoundIdentifier(node.name, (id) => {
+                const { text } = id;
+                if (scope.has(text)) {
+                    ctx.addFailureAtNode(id, Rule.FAILURE_STRING(text));
+                } else {
+                    scope.add(text);
+                }
+            });
         }
 
-        super.visitVariableDeclaration(node);
-    }
-
-    private handleSingleVariableIdentifier(variableIdentifier: ts.Identifier) {
-        const variableName = variableIdentifier.text;
-        const currentBlockScope = this.getCurrentBlockScope();
-
-        if (currentBlockScope.varNames.indexOf(variableName) >= 0) {
-            this.addFailureAtNode(variableIdentifier, Rule.FAILURE_STRING_FACTORY(variableIdentifier.text));
-        } else {
-            currentBlockScope.varNames.push(variableName);
-        }
-    }
+        return ts.forEachChild(node, cb);
+    });
 }
 
-class ScopeInfo {
-    public varNames: string[] = [];
+function forEachBoundIdentifier(name: ts.BindingName, action: (id: ts.Identifier) => void): void {
+    switch (name.kind) {
+        case ts.SyntaxKind.Identifier:
+            return action(name);
+        case ts.SyntaxKind.ObjectBindingPattern:
+            for (const e of name.elements) {
+                forEachBoundIdentifier(e.name, action);
+            }
+            break;
+        case ts.SyntaxKind.ArrayBindingPattern:
```

Comment:
you can make this case clause the `default` clause and handle both binding patterns to avoid duplicated code since both blocks are very similar. you can then drop the `case ObjectBindingPattern`